<?xml version="1.0"?>
<robot name="robotArm" xmlns:xacro="http://www.ros.org/wiki/xacro">

<xacro:include filename="$(find my_arm)/models/pisa_iit_soft_hand/soft_hand.urdf.xacro"/>

<!-- Formula according to https://en.wikipedia.org/wiki/List_of_moment_of_inertia_tensors for cylinder -->
<xacro:macro name="cyl_inertia" params="mass r h ">
    <mass value="${mass}"/>
    <inertia ixx="${0.083333 * mass * (3*r*r + h*h)}" ixy="0" ixz="0"
             iyy="${0.083333 * mass * (3*r*r + h*h)}" iyz="0"
             izz="${0.5*mass*r*r}" />
</xacro:macro>

<!-- Constants for robot dimensions -->
<xacro:property name="PI"           value="3.1415926535897931"/>
<xacro:property name="scale"        value="1.0" /> <!-- scale -->

<xacro:property name="mass"         value="${1 * scale}"     /> <!-- arbitrary value for mass -->
<xacro:property name="mass_divider" value="1.0"              /> <!-- mass divider -->
<xacro:property name="radius_hor_2" value="${1 * radius3}"   />
<xacro:property name="radius_hor_3" value="${1 * radius4}"   />

<xacro:property name="baseLength"   value="${0.2 * scale}"   /> <!-- base_link -->
<xacro:property name="base_radius"  value="${radius2}"       />

<xacro:property name="length2"      value="${0.5 * scale}"   /> <!-- body 2 -->
<xacro:property name="radius2"      value="${0.1 * scale}"   />

<xacro:property name="length20"     value="${2*radius2}"     /> <!-- body 20 -->
<xacro:property name="length3"      value="${0.5 * scale}"   /> <!-- body 3 -->
<xacro:property name="radius3"      value="${0.1 * scale}"   />

<xacro:property name="length30"     value="${2*radius2}"     /> <!-- body 30 -->
<xacro:property name="length4"      value="${0.3 * scale}"   /> <!-- body 4 -->
<xacro:property name="radius4"      value="${0.1 * scale}"   />

<xacro:property name="length5"      value="${0.05 * scale}"  /> <!-- body 5 -->
<xacro:property name="radius5"      value="${0.05 * scale}"  />

<xacro:property name="damping"      value="1" />             <!-- Damping -->
<xacro:property name="axel_offset"  value="0" />             <!-- Space btw top of beam and the each joint -->

<!-- Import all Gazebo-customization elements, including Gazebo colors -->
<xacro:include filename="$(find my_arm)/models/myArm2.gazebo" />
<!-- Import Rviz colors -->
<xacro:include filename="$(find my_arm)/models/materials.xacro" />

<!---............................. Link Definitions ................................-->
<!--
<link name="root"/>

<joint name="root_joint" type="fixed">
    <parent link="root"/>
    <child link="base_link"/>
</joint>
-->

<!---............................. BASE LINK WHITE ................................-->
<link name="base_link">
    <collision> <!-- collision -->
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
            <cylinder length="${baseLength}" radius="${base_radius}"/>
        </geometry>
    </collision>
    <visual>    <!-- visual -->
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
            <cylinder length="${baseLength}" radius="${base_radius}"/>
        </geometry>
        <material name ="white"/>
    </visual>
    <!--
    <inertial>  The root link base_link has an inertia specified in the URDF, but KDL does not support a root link with an inertia.  As a workaround, you can add an extra dummy link to your URDF.
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <xacro:cyl_inertia mass="${mass/mass_divider}" h="${baseLength}" r="${base_radius}"/>
    </inertial>
    -->
</link>

<!---............................. BASE JOINT REVOLUTE Z ................................-->
<joint name="base_joint" type="revolute">
    <parent link="base_link"/>
    <child link="body2"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0 0 ${baseLength/2+length2/2}"/> <!-- z - offset (difference between two joint center) (Base link + Link1) length/2 -->
    <limit effort="1000.0" lower="-2.96" upper="2.96" velocity="0.5"/>
    <dynamics damping="${damping}"/>
</joint>

<!---............................. LINK 2 BLUE ................................-->
<link name="body2">
    <collision> <!-- collision -->
        <origin rpy="0 0 0" xyz="0 0 0"/>
        <geometry>
            <cylinder length="${length2}" radius="${radius2}"/>
        </geometry>
    </collision>
    <visual>    <!-- visual -->
        <origin rpy="0 0 0" xyz="0 0 0"/>
        <geometry>
            <cylinder length="${length2}" radius="${radius2}"/>
        </geometry>
        <material name ="blue"/>
    </visual>
    <inertial>  <!-- inertial -->
        <origin rpy="0 0 0" xyz="0 0 0"/>
        <xacro:cyl_inertia mass="${mass/mass_divider}" h="${length2}" r="${radius2}"/>
    </inertial>
</link>

<!---............................. JOINT 2 CONTINUOUS AROUND Y ................................-->
<joint name="j2" type="continuous">
    <parent link="body2"/>
    <child link="body20"/>
    <origin xyz="0 ${axel_offset} ${length2/2+radius_hor_2}" rpy="0 0 0" />
    <axis xyz="0 1 0"/>
    <limit effort="1000.0" lower="-2.96" upper="2.96" velocity="0.5"/>
    <dynamics damping="${damping}"/>
</joint>

<!---............................. LINK 20 WHITE ................................-->
<link name="body20">
    <collision> <!-- collision -->
        <origin rpy="${PI/2} 0 0" xyz="0 0 0"/> <!-- Due to being rotated around X, z as 0 -->
        <geometry>
            <cylinder length="${length20}" radius="${radius_hor_2}"/> <!-- Double the diameter of link cylinder-->
        </geometry>
    </collision>
    <visual>    <!-- visual -->
        <origin rpy="${PI/2} 0 0" xyz="0 0 0"/>
        <geometry>
            <cylinder length="${length20}" radius="${radius_hor_2}"/> <!-- Double the diameter of link cylinder-->
        </geometry>
        <material name ="white"/>
    </visual>
    <inertial>  <!-- inertial -->
        <origin rpy="0 0 0" xyz="0 0 0"/>
        <xacro:cyl_inertia mass="${mass/mass_divider}" h="${length20}" r="${radius_hor_2}"/>
    </inertial>
</link>

<!---............................. JOINT 2 FIXED ................................-->
<joint name="j20" type="fixed">
    <parent link="body20"/>
    <child link="body3"/>
    <origin xyz="0 ${axel_offset} 0" rpy="0 0 0"/>
    <!-- <axis xyz="0 1 0"/> -->

    <limit effort="1000.0" lower="-2.96" upper="2.96" velocity="0.5"/>
    <dynamics damping="${damping}"/>
</joint>

<!---............................. LINK 3 GREEN .......................................-->
<link name="body3">
    <collision> <!-- collision -->
        <origin rpy="0 0 0" xyz="0 0 ${radius_hor_2+length3/2}"/>
        <geometry>
            <cylinder length="${length3}" radius="${radius3}"/>
        </geometry>
    </collision>
    <visual>    <!-- visual -->
        <origin rpy="0 0 0" xyz="0 0 ${radius_hor_2+length3/2}"/>
        <geometry>
            <cylinder length="${length3}" radius="${radius3}"/>
        </geometry>
        <material name ="green"/>
    </visual>
    <inertial>  <!-- inertial -->
        <origin rpy="0 0 0" xyz="0 0 0"/>
        <xacro:cyl_inertia mass="${mass/mass_divider}" h="${length3}" r="${radius3}"/>
    </inertial>
</link>

<!---............................. JOINT 20 CONTINUOUS AROUND Y ................................-->
<joint name="j3" type="continuous">
    <parent link="body3"/>
    <child link="body30"/>
    <origin xyz="0 ${axel_offset} ${radius_hor_2 + length3 + radius_hor_3}" rpy="0 0 0" />
    <axis xyz="0 1 0"/>
    <limit effort="1000.0" lower="-2.96" upper="2.96" velocity="0.5"/>
    <dynamics damping="${damping}"/>
</joint>

<!---............................. LINK 30 WHITE ................................-->
<link name="body30">
    <collision> <!-- collision -->
        <origin rpy="${PI/2} 0 0" xyz="0 0 0"/>
        <geometry>
            <cylinder length="${length30}" radius="${radius_hor_3}"/> <!-- Double the diameter of link cylinder-->
        </geometry>
    </collision>
    <visual>    <!-- visual -->
        <origin rpy="${PI/2} 0 0" xyz="0 0 0"/>
        <geometry>
            <cylinder length="${length30}" radius="${radius_hor_3}"/> <!-- Double the diameter of link cylinder-->
        </geometry>
        <material name ="white"/>
    </visual>
    <inertial>  <!-- inertial -->
        <origin rpy="0 0 0" xyz="0 0 0"/>
        <xacro:cyl_inertia mass="${mass/mass_divider}" h="${length30}" r="${radius_hor_3}"/>
    </inertial>
</link>

<!---............................. JOINT 3 FIXED ................................-->
<joint name="j30" type="fixed">
    <parent link="body30"/>
    <child link="body4"/>
    <origin xyz="0 ${axel_offset} 0" rpy="0 0 0"/>
    <!--<axis xyz="0 1 0"/> -->

    <limit effort="1000.0" lower="-2.96" upper="2.96" velocity="0.5"/>
    <dynamics damping="${damping}"/>
</joint>

<!---............................. LINK 4 GREEN .......................................-->
<link name="body4">
    <collision> <!-- collision -->
        <origin rpy="0 0 0" xyz="0 0 ${radius_hor_3+length4/2}"/>
        <geometry>
            <cylinder length="${length4}" radius="${radius4}"/>
        </geometry>
    </collision>
    <visual>    <!-- visual -->
        <origin rpy="0 0 0" xyz="0 0 ${radius_hor_3+length4/2}"/>
        <geometry>
            <cylinder length="${length4}" radius="${radius4}"/>
        </geometry>
        <material name ="green"/>
    </visual>
    <inertial>  <!-- inertial -->
        <origin rpy="0 0 0" xyz="0 0 0"/>
        <xacro:cyl_inertia mass="${mass/mass_divider}" h="${length4}" r="${radius4}"/>
    </inertial>
</link>

<!---............................. JOINT 4 REVOLUTE Z ................................ -->
<joint name="j4" type="revolute">
    <parent link="body4"/>
    <child link="softHandWrist"/>
    <origin xyz="0 ${axel_offset} ${radius_hor_3+length4}" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
    <limit effort="1000.0" lower="-2.96" upper="2.96" velocity="0.5"/>
    <dynamics damping="${2 * damping}"/>
</joint>

<!---............................. HAND .............................................. -->

<link name="softHandWrist">
    <collision> <!-- collision -->
        <origin rpy="0 0 0" xyz="0 0 ${length5/2}"/>
        <geometry>
            <cylinder length="${length5}" radius="${radius5}"/>
        </geometry>
    </collision>
    <visual>    <!-- visual -->
        <origin rpy="0 0 0" xyz="0 0 ${length5/2}"/>
        <geometry>
            <cylinder length="${length5}" radius="${radius5}"/>
        </geometry>
        <material name ="SoftHand/LightGrey"/>
    </visual>
    <inertial>  <!-- inertial -->
        <origin rpy="0 0 0" xyz="0 0 0"/>
        <xacro:cyl_inertia mass="${mass/mass_divider}" h="${length5}" r="${radius5}"/>
    </inertial>
</link>

<xacro:soft_hand name="softHand" parent="softHandWrist"
                 withAdaptiveTransmission="true" useMimicTag="false" left="true">
  <origin xyz="0 0 ${length5/2}"
          rpy="0 0 0"/>
</xacro:soft_hand>
</robot>
